machinelearning/llama.cpp.md

```diff
@@ -27,46 +27,79 @@ aws ec2 describe-images --owners amazon --filters "Name=architecture,Values=arm6
+**Using llama-cli**
 
-# Now, launch llama-cli with the above model and a sample input prompt. Use Q4_0_8_8 layout model if running on
-# Graviton3 or use the Q4_0_4_8 layout model if running on Graviton4 based EC2 Instances. The following command is using 64 threads.
+```
+# Now, launch llama-cli with the above model and a sample input prompt. The following command is using 64 threads.
 # Change -t argument for running inference with lower thread count. On completion, the script prints throughput and latency metics
 # for prompt encoding and response generation.
-./llama-cli -m meta-llama-3-8b-instruct.Q4_0_8_8.gguf -p "Building a visually appealing website can be done in ten simple steps:" -n 512 -t 64
-
+./build/bin/llama-cli -m meta-llama-3-8b-instruct.Q4_0.gguf -p "Building a visually appealing website can be done in ten simple steps:" -n 512 -t 64
 
 # Launch the model in conversation (chatbot) mode using this command
-./llama-cli -m meta-llama-3-8b-instruct.Q4_0_8_8.gguf -p "You are a helpful assistant" -cnv --color
+./build/bin/llama-cli -m meta-llama-3-8b-instruct.Q4_0.gguf -p "You are a helpful assistant" -cnv --color
+
+```
+
+**Using llama.cpp python binding**
+
+Note: Set the `n_threads` to number of vcpus explicitly while creating the Llama object. This is required to use all cores(vcpus) on Graviton instances. Without this set, the python bindings use half of the vcpus and the performance is not the best.
+
+```
+import json
+import argparse
+
+from llama_cpp import Llama
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-m", "--model", type=str, default="../models/7B/ggml-models.bin")
+args = parser.parse_args()
+
+# for example, for a .16xlarge instance, set n_threads=64
+llm = Llama(model_path=args.model,
+            n_threads=64)
+
+output = llm(
+"Question: How to build a visually appealing website in ten steps? Answer: ",
+max_tokens=512,
+echo=True,
+)
 
 ```
 
 # Additional Resources
 
 Please refer to
 1. [Best-in-class LLM performance on Arm Neoverse V1 based AWS Graviton3 CPUs](https://community.arm.com/arm-community-blogs/b/infrastructure-solutions-blog/posts/best-in-class-llm-performance) to know the LLM inference performance measured on AWS Graviton3 based EC2 Instances.
-2. [Run LLMs on CPU with Amazon SageMaker Real-time Inference](https://community.aws/content/2eazHYzSfcY9flCGKsuGjpwqq1B/run-llms-on-cpu-with-amazon-sagemaker-real-time-inference?lang=en) for running LLMs for real-time inference using AWS Graviton3 and Amazon SageMaker.
+2. [Running Llama 3 70B on the AWS Graviton4 CPU with Human Readable Performance](https://community.arm.com/arm-community-blogs/b/infrastructure-solutions-blog/posts/running-llama-3-70b-on-aws-graviton4) for LLM inference performance on AWS Graviton4 based EC2 Instances.
+3. [Intro to Llama on Graviton](https://dev.to/aws-heroes/intro-to-llama-on-graviton-1dc) for a step by step guide on how to deploy an LLM model on AWS Graviton-based EC2 Instances.
```

Comment:
This page has quantization that is specific to each version of Graviton. Would you still want to leave this reference in here? Maybe add a note that quantization isn't applicable with the latest `llama` release.
```
./llama-quantize --allow-requantize dolphin-2.9.4-llama3.1-8b-Q4_0.gguf dolphin-2.9.4-llama3.1-8b-Q4_0_8_8.gguf Q4_0_8_8
```

Reply:
good point, let me add.